Framework/Geometry/src/Objects/Object.cpp

```diff
@@ -1475,6 +1476,108 @@ double Object::ConeSolidAngle(const V3D &observer,
 }
 
 /**
+ * For simple shapes, the volume is calculated exactly. For more
+ * complex cases, we fall back to Monte Carlo.
+ * @return The volume.
+ */
+double Object::volume() const {
+  int type;
+  double height;
+  double radius;
+  std::vector<Kernel::V3D> vectors;
+  this->GetObjectGeom(type, vectors, radius, height);
+  GluGeometryHandler::GeometryType gluType =
+      static_cast<GluGeometryHandler::GeometryType>(type);
+  switch (gluType) {
+  case GluGeometryHandler::GeometryType::CUBOID: {
+    double volume = 0.0;
+    const Kernel::V3D vertex12 = vectors[1] + vectors[2];
+    const Kernel::V3D vertex13 = vectors[1] + vectors[3];
+    const Kernel::V3D vertex23 = vectors[2] + vectors[3];
+    const Kernel::V3D vertex123 = vectors[1] + vertex23;
+    volume += vertex23.scalar_prod(vectors[3].cross_prod(vectors[2]));
+    volume += vertex23.scalar_prod(vectors[2].cross_prod(vertex123));
+    volume += vertex23.scalar_prod(vertex123.cross_prod(vectors[3]));
+    volume += vertex12.scalar_prod(vectors[2].cross_prod(vectors[1]));
+    volume += vertex12.scalar_prod(vectors[1].cross_prod(vertex13));
+    volume += vertex12.scalar_prod(vertex123.cross_prod(vectors[2]));
+    volume += vectors[1].scalar_prod(vectors[3].cross_prod(vertex13));
+    volume += vertex13.scalar_prod(vectors[3].cross_prod(vertex123));
+    volume += vertex13.scalar_prod(vertex123.cross_prod(vertex23));
+    return volume / 6;
+  }
+  case GluGeometryHandler::GeometryType::SPHERE:
+    return 4.0 / 3.0 * M_PI * radius * radius * radius;
+  case GluGeometryHandler::GeometryType::CYLINDER:
+    return M_PI * radius * radius * height;
+  default:
+    // Fall back to Monte Carlo method.
+    return monteCarloVolume();
+  }
+}
+
+/**
+ * The volume is calculated using the Monte Carlo method.
+ * @returns The volume of this object.
+ */
+double Object::monteCarloVolume() const {
+  const auto &boundingBox = getBoundingBox();
+  if (boundingBox.isNull()) {
+    return 0;
+  }
+  int total = 0;
+  int totalHits = 0;
+  // The maximum number of iterations is chosen by the
+  // Stetson-Harrison method. This means that the following
+  // number may or may not work in every scenario.
+  const int maxIterations = 1000000;
+  const double boundingDx = boundingBox.xMax() - boundingBox.xMin();
+  const double boundingDy = boundingBox.yMax() - boundingBox.yMin();
+  const double boundingDz = boundingBox.yMax() - boundingBox.yMin();
+  PARALLEL {
+    const auto threadCount = PARALLEL_NUMBER_OF_THREADS;
+    const auto currentThreadNum = PARALLEL_THREAD_NUMBER;
+    size_t blocksize = maxIterations / threadCount;
+    if (currentThreadNum == threadCount - 1) {
+      // Last thread may do some extra work.
+      blocksize = maxIterations - (threadCount - 1) * blocksize;
+    }
```

Comment:
In the worst case, the last thread could have significantly more iterations as any other thread. It would be nice to see these split up blocksize never differs by more than 1.

Rounding up `maxIterations / threadCount` would be the simplest option, but may add a few extra iterations.